### PR TITLE
[Falcon] Set `use_cache=False` before creating `presents` which relies on `use_cache`

### LINKED ISSUE #26327: [Falcon] forward pass will fail if `use_cache` is automatically flipped to False
### System Info

- `transformers` version: 4.33.2
- Platform: Linux-5.15.111.1-rolling-lts-linkedin-x86_64-with-glibc2.17
- Python version: 3.10.2
- Huggingface_hub version: 0.14.1
- Safetensors version: 0.3.1
- Accelerate version: 0.21.0
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.1.0a0+gitf998869 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: 8 A100 GPUs
- Using distributed or parallel set-up in script?: Nah just `torchrun`

### Who can help?

@ArthurZucker and @younesbelkada

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

Full-parameter fine-tune a Falcon 180B using any kind of task, using activation checkpointing (`--gradient_checkpointing`).
 
When doing `model = transformers.AutoModelForCausalLM.from_pretrained()`, don't set `use_cache=False` but leave it default.

This will result in the `use_cache` flag being [flapped to False](https://github.com/huggingface/transformers/blob/39df4eca739b0870f73dbcfdfa09179e3135c75d/src/transformers/models/falcon/modeling_falcon.py#L951). But the `presents` used for cache is not reset to `None` and later on [this code branch](https://github.com/huggingface/transformers/blob/39df4eca739b0870f73dbcfdfa09179e3135c75d/src/transformers/models/falcon/modeling_falcon.py#L993) which should be exclusive for `use_cache=True` will be entered and then hit following error:
```
  File "/home/jobuser/.local/lib/python3.10/site-packages/accelerate/utils/operations.py", line 636, in forward
    return model_forward(*args, **kwargs)
  File "/home/jobuser/.local/lib/python3.10/site-packages/accelerate/utils/operations.py", line 624, in __call__
    return convert_to_fp32(self.model_forward(*args, **kwargs))
  File "/home/jobuser/.local/lib/python3.10/site-packages/torch/amp/autocast_mode.py", line 16, in decorate_autocast
    return func(*args, **kwargs)
  File "/home/jobuser/.local/lib/python3.10/site-packages/transformers/models/falcon/modeling_falcon.py", line 1031, in forward
    transformer_outputs = self.transformer(
  File "/home/jobuser/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/jobuser/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/jobuser/.local/lib/python3.10/site-packages/transformers/models/falcon/modeling_falcon.py", line 952, in forward
    presents = self._convert_cache_to_standard_format(presents, batch_size)
  File "/home/jobuser/.local/lib/python3.10/site-packages/transformers/models/falcon/modeling_falcon.py", line 705, in _convert_cache_to_standard_format
    batch_size_times_num_heads, kv_length, head_dim = past_key_value[0][0].shape
IndexError: tuple index out of range
```

### Expected behavior

The `presents` tuple needs to be set to None along with `use_cache=False`

### PR DESCRIPTION
**Pull request by yundai424**
# What does this PR do?

Fixes #26327 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

- text models: @ArthurZucker and @younesbelkada

### CODE REVIEW

File: src/transformers/models/falcon/modeling_falcon.py [hunk 1]
```diff
@@ -949,6 +949,7 @@ def forward(
                         "`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`..."
                     )
                     use_cache = False
+                    presents = None
```
Comment (ArthurZucker) [inline src/transformers/models/falcon/modeling_falcon.py:952]: This should be done outside the loop and similar to this:
```python
        if self.gradient_checkpointing and self.training:
            if use_cache:
                logger.warning_once(
                    "`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`..."
                )
                use_cache = False

        present_key_values = () if use_cache else None
        all_self_attentions = () if output_attentions else None
        all_hidden_states = () if output_hidden_states else None
```

**Review by ArthurZucker — COMMENTED** [review]
That's a good catch!

File: src/transformers/models/falcon/modeling_falcon.py (hunk 1 above)
Comment (yundai424) [reply src/transformers/models/falcon/modeling_falcon.py]: updated, thanks!

**Review by younesbelkada — COMMENTED** [review]
Great catch ! Can you run the styling checks? 
```bash
make fixup
```
Then we can merge I think

**Comment by younesbelkada** [comment]
Hi @yundai424 thanks a lot for iterating, in order to move forward with the PR could you merge your branch with main branch?

**Comment by younesbelkada** [comment]
Hi @yundai424 
I meant to merge your local branch with HF's main branch, assuming the hf remote is tagged as `upstream` (you can also add it with `git remote add upstream https://github.com/huggingface/transformers.git`)

```bash
git fetch upstream
git merge upstream/main
git push
```

**Comment by yundai424** [comment]
oh cool i see what you mean.. merged, thanks! @younesbelkada

**Review by younesbelkada — APPROVED** [review]
Clean to me, thanks!

**Review by LysandreJik — APPROVED** [review]
Great, thanks @yundai424!

**Pull request merged**